Improve witness needed logic

lehins opened the pull request:
# Description

## Abstract deciding which certificates require witnesses.

Certificates structure has changed in Conway, because of that old logic,
even with backwards compatibility translation, will not work correctly.
In particular, some new certificates did not require a witness, where
they should have required one. Abstracting the logic by adding
`getVKeyWitnessTxCert` and `getScriptWitnessTxCert` to `EraTxCert` we
can solve this problem with different implemnetation on per era basis.
This also opens up an opportunitiy for deduplication implemented in
a subsequent commit.

## Make registration certificates with deposits require witnesses

It was discusses during a meeting that it was an unfortunate requirement
in Shelley to not enforce a witness for staking credential registration.
It was implemented to allow ITIN rewards, which is not a feature anymore.
Therefore this is no longer a necessary requirement and it was decided
to change this by enforcing a wtiness for registration certificates, but
only for the ones that contain a deposit. This would mean that we can
keep backwards compatibility throughout Conway, but in the next era it
will become a proper requirement.

## Reduce duplication by consolidatiing two versions of witsVKeyNeeded

Implementations in Shelley and Alonzo were almost identical.

## Avoid usage of partial functions

In three different places `error` call was removed by improving the logic.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Any changes are noted in the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins committed:
Abstract deciding which certificates require witnesses.

Certificates structure has changed in Conway, because of that old logic,
even with backwards compatibility translation, will not work correctly.
In particular, some new certificates did not require a witness, where
they should have required one. Abstracting the logic by adding
`getVKeyWitnessTxCert` and `getScriptWitnessTxCert` to `EraTxCert` we
can solve this problem with different implemnetation on per era basis.
This also opens up an opportunitiy for deduplication implemented in
a subsequent commit.

lehins committed:
Reduce duplication by consolidatiing two versions of `witsVKeyNeeded`

Implementations in Shelley and Alonzo were almost identical.

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs:
```diff
@@ -152,22 +146,11 @@ getAlonzoScriptsNeeded (UTxO u) txBody =
         withdrawals = unWithdrawals $ txBody ^. withdrawalsTxBodyL
         fromRwd accnt = do
-          hash <- scriptCred $ getRwdCred accnt
+          hash <- credScriptHash $ getRwdCred accnt
```
how about `scriptCredHash`? since it is the hash of a script credential.

lehins replied:
That's right, that's why `credScriptHash` is better, because it is a `ScriptHash`, not a `CredHash` :smile: 

Also this way it is consistent with `credKeyHashWitness`, which is quite similar, but for the `KeyHash`es

JaredCorduan replied:
:laughing: yep, makes sense now.

JaredCorduan commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs:
```diff
@@ -454,14 +447,14 @@ validateVerifiedWits tx =
-validateNeededWitnesses witsvkeyneeded genDelegs utxo tx witsKeyHashes =
-  let needed = witsvkeyneeded utxo tx genDelegs
+validateNeededWitnesses genDelegs utxo tx witsKeyHashes =
+  let needed = witsVKeyNeeded utxo tx genDelegs
```
nice reduction!

JaredCorduan commented:
nice clean up!

lehins committed:
Make registration certificates with deposits require witnesses

It was discusses during a meeting that it was an unfortunate requirement
in Shelley to not enforce a witness for staking credential registration.
It was implemented to allow ITIN rewards, which is not a feature anymore.
Therefore this is no longer a necessary requirement and it was decided
to change this by enforcing a wtiness for registration certificates, but
only for the ones that contain a deposit. This would mean that we can
keep backwards compatibility throughout Conway, but in the next era it
will become a proper requirement.

lehins committed:
Deprecate `propWits`. It will become an internal function

JaredCorduan commented on eras/mary/impl/src/Cardano/Ledger/Mary/TxCert.hs:
```diff
@@ -5,13 +5,17 @@ module Cardano.Ledger.Mary.TxCert () where
 import Cardano.Ledger.Crypto (Crypto, StandardCrypto)
 import Cardano.Ledger.Mary.Era (MaryEra)
-import Cardano.Ledger.Shelley.TxCert (EraTxCert (..), ShelleyEraTxCert (..), ShelleyTxCert (..))
+import Cardano.Ledger.Shelley.TxCert
```
non-Core open import?

lehins replied:
There are three imports in this module. Only one of them is open, which means that everything that is not imported explicitly comes from `Cardano.Ledger.Shelley.TxCert` (aside from Prelude of course)

That being said, I know how much you like those explicit imports, so I'll add it back.

JaredCorduan replied:
It's true I don't like them. but more I was just hoping for a consistent rule about when we use them. I can be flexible too :)

JaredCorduan reviewed (approved):
Looks great, much cleaner and safer!
